cpp/examples/arrow/execution_plan_documentation_examples.cc

```diff
@@ -0,0 +1,1160 @@
+    arrow::dataset::internal::Initialize();
+
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<cp::ExecPlan> plan,
+                            cp::ExecPlan::Make(&exec_context));
+
+    std::shared_ptr<arrow::dataset::Dataset> dataset = CreateDataset();
+
+    auto options = std::make_shared<arrow::dataset::ScanOptions>();
+    // sync scanning is not supported by ScanNode
+    options->use_async = true;
+    // specify the filter
+    cp::Expression b_is_true = cp::field_ref("b");
+    options->filter = b_is_true;
+    // for now, specify the projection as the full project expression (eventually this can
+    // just be a list of materialized field names)
+
+    cp::Expression a_times_2 = cp::call("multiply", {cp::field_ref("a"), cp::literal(2)});
+    options->projection =
+        cp::call("make_struct", {a_times_2}, cp::MakeStructOptions{{"a * 2"}});
+
+    // // construct the scan node
+    cp::ExecNode* scan;
+
+    auto scan_node_options = arrow::dataset::ScanNodeOptions{dataset, options};
+
+    ARROW_ASSIGN_OR_RAISE(scan,
+                            cp::MakeExecNode("scan", plan.get(), {}, scan_node_options));
+
+    // pipe the scan node into a filter node
+    cp::ExecNode* filter;
+    ARROW_ASSIGN_OR_RAISE(filter, cp::MakeExecNode("filter", plan.get(), {scan},
+                                                    cp::FilterNodeOptions{b_is_true}));
+
+    cp::ExecNode* project;
+
+    ARROW_ASSIGN_OR_RAISE(project,
+                            cp::MakeExecNode("augmented_project", plan.get(), {filter},
+                                            cp::ProjectNodeOptions{{a_times_2}}));
+
+    // // finally, pipe the project node into a sink node
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+    ARROW_ASSIGN_OR_RAISE(cp::ExecNode * sink,
+                            cp::MakeExecNode("sink",
+                            plan.get(), {project},
+                            cp::SinkNodeOptions{&sink_gen}));
+
+    ABORT_ON_FAILURE(sink->Validate());
+
+    // // translate sink_gen (async) to sink_reader (sync)
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader =
+        cp::MakeGeneratorReader(arrow::schema({arrow::field("a * 2", arrow::int32())}),
+                                std::move(sink_gen), exec_context.memory_pool());
+
+    // // validate the plan
+    ABORT_ON_FAILURE(plan->Validate());
+    PRINT_LINE("Exec Plan created: " << plan->ToString());
+    // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+
+    // // collect sink_reader into a Table
+    std::shared_ptr<arrow::Table> response_table;
+    ARROW_ASSIGN_OR_RAISE(response_table,
+                            arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+    std::cout << "Results : " << response_table->ToString() << std::endl;
+
+    // // stop producing
+    plan->StopProducing();
+
+    // // plan mark finished
+    plan->finished().Wait();
+
+    return arrow::Status::OK();
```

Comment:
We should return the status from `plan->finished()` to catch if the plan errors.